File: frontend/src/services/maintenance-dashboard.service.ts
/**
 * Maintenance Dashboard API Service
 * Story 8.4: Maintenance Dashboard
 */

import { apiClient } from '@/lib/api';
import type {
  MaintenanceDashboard,
  JobsByStatus,
  JobsByPriority,
  JobsByCategory,
  HighPriorityJobsPage,
  RecentlyCompletedJob,
  MaintenanceDashboardFilters,
  HighPriorityJobsFilters
} from '@/types/maintenance-dashboard';

const MAINTENANCE_DASHBOARD_URL = '/v1/dashboard/maintenance';

/**
 * Helper to build query params
 */
const buildParams = (filters: Record<string, unknown>): URLSearchParams => {
  const params = new URLSearchParams();
  Object.entries(filters).forEach(([key, value]) => {
    if (value !== undefined && value !== null && value !== '') {
      params.append(key, String(value));
    }
  });
  return params;
};

/**
 * Maintenance Dashboard Service
 * Provides API client methods for all maintenance dashboard endpoints
 */
export const maintenanceDashboardService = {
  /**
   * Get complete maintenance dashboard data (AC-10)
   * GET /api/v1/dashboard/maintenance
   */
  async getMaintenanceDashboard(
    filters?: MaintenanceDashboardFilters
  ): Promise<MaintenanceDashboard> {
    const params = buildParams((filters || {}) as Record<string, unknown>);
    const response = await apiClient.get<MaintenanceDashboard>(
      `${MAINTENANCE_DASHBOARD_URL}?${params.toString()}`
    );
    return response.data;
  },

  /**
   * Get jobs by status for pie chart (AC-11)
   * GET /api/v1/dashboard/maintenance/jobs-by-status
   */
  async getJobsByStatus(
    filters?: MaintenanceDashboardFilters
  ): Promise<JobsByStatus[]> {
    const params = buildParams((filters || {}) as Record<string, unknown>);
    const response = await apiClient.get<JobsByStatus[]>(
      `${MAINTENANCE_DASHBOARD_URL}/jobs-by-status?${params.toString()}`
    );
    return response.data;
  },

  /**
   * Get jobs by priority for bar chart (AC-12)
   * GET /api/v1/dashboard/maintenance/jobs-by-priority
   */
  async getJobsByPriority(
    filters?: MaintenanceDashboardFilters
  ): Promise<JobsByPriority[]> {
    const params = buildParams((filters || {}) as Record<string, unknown>);
    const response = await apiClient.get<JobsByPriority[]>(
      `${MAINTENANCE_DASHBOARD_URL}/jobs-by-priority?${params.toString()}`
    );
    return response.data;
  },

  /**
   * Get jobs by category for horizontal bar chart (AC-13)
   * GET /api/v1/dashboard/maintenance/jobs-by-category
   */
  async getJobsByCategory(
    filters?: MaintenanceDashboardFilters
  ): Promise<JobsByCategory[]> {
    const params = buildParams((filters || {}) as Record<string, unknown>);
    const response = await apiClient.get<JobsByCategory[]>(
      `${MAINTENANCE_DASHBOARD_URL}/jobs-by-category?${params.toString()}`
    );
    return response.data;
  },

  /**
   * Get high priority and overdue jobs with pagination (AC-14)
   * GET /api/v1/dashboard/maintenance/high-priority-overdue
   */
  async getHighPriorityOverdueJobs(
    filters?: HighPriorityJobsFilters
  ): Promise<HighPriorityJobsPage> {
    const params = buildParams({
      propertyId: filters?.propertyId,
      status: filters?.status,
      page: filters?.page ?? 0,
      size: filters?.size ?? 10
    });
    const response = await apiClient.get<HighPriorityJobsPage>(
      `${MAINTENANCE_DASHBOARD_URL}/high-priority-overdue?${params.toString()}`
    );
    return response.data;
  },

  /**
   * Get recently completed jobs (AC-15)
   * GET /api/v1/dashboard/maintenance/recently-completed
   */
  async getRecentlyCompletedJobs(
    filters?: MaintenanceDashboardFilters,
    limit: number = 5
  ): Promise<RecentlyCompletedJob[]> {
    const params = buildParams({
      ...filters,
      limit
    });
    const response = await apiClient.get<RecentlyCompletedJob[]>(
      `${MAINTENANCE_DASHBOARD_URL}/recently-completed?${params.toString()}`
    );
    return response.data;
  }
};

export default maintenanceDashboardService;

File: backend/src/main/java/com/ultrabms/service/MaintenanceDashboardService.java
package com.ultrabms.service;

import com.ultrabms.dto.dashboard.maintenance.*;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;
import java.util.UUID;

/**
 * Service interface for Maintenance Dashboard operations.
 * Provides KPI calculations, chart data, and job lists.
 *
 * Story 8.4: Maintenance Dashboard
 */
public interface MaintenanceDashboardService {

    // =================================================================
    // COMPLETE DASHBOARD (AC-10)
    // =================================================================

    /**
     * Get complete maintenance dashboard data in a single call
     * AC-10: GET /api/v1/dashboard/maintenance endpoint
     *
     * @param propertyId Optional property filter (null for all properties)
     * @return Complete maintenance dashboard data
     */
    MaintenanceDashboardDto getMaintenanceDashboard(UUID propertyId);

    // =================================================================
    // CHART DATA ENDPOINTS
    // =================================================================

    /**
     * Get jobs grouped by status for pie chart
     * AC-11: GET /api/v1/dashboard/maintenance/jobs-by-status endpoint
     *
     * @param propertyId Optional property filter
     * @return List of jobs by status for chart
     */
    List<JobsByStatusDto> getJobsByStatus(UUID propertyId);

    /**
     * Get jobs grouped by priority for bar chart
     * AC-12: GET /api/v1/dashboard/maintenance/jobs-by-priority endpoint
     *
     * @param propertyId Optional property filter
     * @return List of jobs by priority for chart
     */
    List<JobsByPriorityDto> getJobsByPriority(UUID propertyId);

    /**
     * Get jobs grouped by category for horizontal bar chart
     * AC-13: GET /api/v1/dashboard/maintenance/jobs-by-category endpoint
     *
     * @param propertyId Optional property filter
     * @return List of jobs by category for chart (sorted by count desc)
     */
    List<JobsByCategoryDto> getJobsByCategory(UUID propertyId);

    // =================================================================
    // LIST DATA ENDPOINTS
    // =================================================================

    /**
     * Get high priority and overdue jobs with pagination
     * AC-14: GET /api/v1/dashboard/maintenance/high-priority-overdue endpoint
     *
     * @param propertyId Optional property filter
     * @param statusFilter Optional status filter for click-to-filter
     * @param pageable Pagination parameters
     * @return Page of high priority and overdue jobs
     */
    Page<HighPriorityJobDto> getHighPriorityAndOverdueJobs(UUID propertyId, String statusFilter, Pageable pageable);

    /**
     * Get recently completed jobs
     * AC-15: GET /api/v1/dashboard/maintenance/recently-completed endpoint
     *
     * @param propertyId Optional property filter
     * @param limit Maximum items to return (default 5)
     * @return List of recently completed jobs
     */
    List<RecentlyCompletedJobDto> getRecentlyCompletedJobs(UUID propertyId, int limit);
}
